engine/pkg/rpcerror/attributes.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcerror
+
+import "google.golang.org/grpc/codes"
+
+type retryablity interface {
+	isRetryable() bool
+}
+
+// Retryable can be passed as a parameter to Error to
+// mark the error as retryable.
+type Retryable struct{}
+
+func (r Retryable) isRetryable() bool {
+	return true
+}
+
+// NotRetryable can be passed as a parameter to Error to
+// mark the error as not retryable.
+type NotRetryable struct{}
+
+func (n NotRetryable) isRetryable() bool {
+	return false
+}
+
+type grpcStatusCoder interface {
+	grpcStatusCode() codes.Code
+}
+
+// Canceled can be passed as a parameter to Error to
+// mark the grpc error code of this error as Canceled.
+type Canceled struct{}
```

Comment:
Why do we define so many types? Can we define the errors as an enum type? If so, it's easy to implement the `grpcStatusCode` method.

Reply:
Because types can be used as parameters to `Error[E, R]`.

Reply:
Why not just put `statusCode` and `retryable` as `Error`'s member variables directly?

Reply:
> Why not just put `statusCode` and `retryable` as `Error`'s member variables directly?

It's not convenient for defining an error type with a type definition only. For example

```go
type ErrMyCustomError struct {
    rpcerror.Error[rpcerror.Retryable, rpcerror.Unavailable]

    field1 int
    field2 string
}
```